### ISSUE TITLE
Timeout causing re-registration

### ISSUE THREAD

**Issue by stongo** [coreos/etcd#3257]
version: 2.0.12
coreos channel: beta

When creating new clusters, I've been getting this error on and off all day. Seems like etcd2 is prematurely exiting on a timeout and then trying to re-register again.

```
-- Logs begin at Mon 2015-08-10 22:36:40 UTC, end at Mon 2015-08-10 23:00:53 UTC. --
Aug 10 22:36:51 host-00005 systemd[1]: Started etcd2.
Aug 10 22:36:51 host-00005 systemd[1]: Starting etcd2...
Aug 10 22:36:51 host-00005 etcd2[1027]: 2015/08/10 22:36:51 etcd: listening for peers on http://10.210.5.99:2380
Aug 10 22:36:51 host-00005 etcd2[1027]: 2015/08/10 22:36:51 etcd: listening for client requests on http://0.0.0.0:2379
Aug 10 22:36:51 host-00005 etcd2[1027]: 2015/08/10 22:36:51 etcd: listening for client requests on http://0.0.0.0:4001
Aug 10 22:36:51 host-00005 etcd2[1027]: 2015/08/10 22:36:51 etcdserver: datadir is valid for the 2.0.1 format
Aug 10 22:36:52 host-00005 etcd2[1027]: 2015/08/10 22:36:52 discovery: found peer a2efc57d58 in the cluster
Aug 10 22:36:52 host-00005 etcd2[1027]: 2015/08/10 22:36:52 discovery: found self 2ec1d9da0e in the cluster
Aug 10 22:36:52 host-00005 etcd2[1027]: 2015/08/10 22:36:52 discovery: found 2 peer(s), waiting for 1 more
Aug 10 22:37:35 host-00005 etcd2[1027]: 2015/08/10 22:37:35 discovery: during waiting for other nodes connection to https://discovery.etcd.io timed out, retrying in 2s
Aug 10 22:37:37 host-00005 systemd[1]: etcd2.service: Main process exited, code=exited, status=1/FAILURE
Aug 10 22:37:37 host-00005 systemd[1]: etcd2.service: Unit entered failed state.
Aug 10 22:37:37 host-00005 systemd[1]: etcd2.service: Failed with result 'exit-code'.
Aug 10 22:37:47 host-00005 systemd[1]: etcd2.service: Service hold-off time over, scheduling restart.
Aug 10 22:37:47 host-00005 systemd[1]: Started etcd2.
Aug 10 22:37:47 host-00005 systemd[1]: Starting etcd2...
Aug 10 22:37:47 host-00005 etcd2[1333]: 2015/08/10 22:37:47 etcd: listening for peers on http://10.210.5.90:2380
Aug 10 22:37:47 host-00005 etcd2[1333]: 2015/08/10 22:37:47 etcd: listening for client requests on http://0.0.0.0:2379
Aug 10 22:37:47 host-00005 etcd2[1333]: 2015/08/10 22:37:47 etcd: listening for client requests on http://0.0.0.0:4001
Aug 10 22:37:47 host-00005 etcd2[1333]: 2015/08/10 22:37:47 etcdserver: datadir is valid for the 2.0.1 format
Aug 10 22:37:48 host-00005 etcd2[1333]: 2015/08/10 22:37:48 etcd: stopping listening for client requests on http://0.0.0.0:4001
Aug 10 22:37:48 host-00005 etcd2[1333]: 2015/08/10 22:37:48 etcd: stopping listening for client requests on http://0.0.0.0:2379
Aug 10 22:37:48 host-00005 etcd2[1333]: 2015/08/10 22:37:48 etcd: stopping listening for peers on http://10.210.5.90:2380
Aug 10 22:37:48 host-00005 etcd2[1333]: 2015/08/10 22:37:48 member "9d33e0f5aa27" has previously registered with discovery service token (https://discovery.etcd.io/0d3766
Aug 10 22:37:48 host-00005 etcd2[1333]: 2015/08/10 22:37:48 But etcd could not find vaild cluster configuration in the given data dir (/var/lib/etcd2).
Aug 10 22:37:48 host-00005 etcd2[1333]: 2015/08/10 22:37:48 Please check the given data dir path if the previous bootstrap succeeded
Aug 10 22:37:48 host-00005 etcd2[1333]: 2015/08/10 22:37:48 or use a new discovery token if the previous bootstrap failed.
```

**Comment by yichengq** [comment]
discovery is trying forever now due to https://github.com/coreos/etcd/issues/3012. As long as process is not killed, this will not be a problem.

**Comment by xiang90** [comment]
@yichengq Can we port that back to 2.1?

**Comment by yichengq** [comment]
I don't want to add ClusterError-related commits to 2.1 because it includes lots of commits. But I could write a simple one to let it retry on client errors.

**Comment by yichengq** [comment]
After attempt, i find it is hard to have the same functionality easily, so I would backport related commits back to release-2.1, and it is supposed to be in 2.1.2.

**Comment by bobelev** [comment]
I am experiencing same issue. How can I start cluster now? One of my nodes always can't join within timeout.

**Comment by yichengq** [comment]
@bobelev You could try static bootstrap here: https://github.com/coreos/etcd/blob/master/Documentation/clustering.md#static

**Comment by stongo** [comment]
You can also delete the failing member from the discovery url and restart
etcd2 on machine having trouble

**Comment by xiang90** [comment]
@stongo Please try not to do so. A better way is to generate a new token.

**Comment by bobelev** [comment]
Thank you.
I started etcd2 on all machines in non-daemon mode, waited for them to register and relaunched them as a service.

**Comment by xiang90** [comment]
@stongo @bobelev We fixed this in the recent 2.1.2 release. Hope you can give it a try. Thanks!
